### PR TITLE
SPHINCS+

### PR DESCRIPTION
**Pull request by reneme**
#3099

This is a huge mess at the moment... no need to look deeper into it at this point.

## TODO

* [ ] Decide which KAT vectors to use and whether to disable some vectors due to timing constraints. The full KAT body would likely take way to long.

## Implementation Milestones

That's a rough collection of implementation work packages. We'll keep it updated as things progress.

* [x] FORS signatures
* [x] WOTS signatures
* [x] XMSS-ish trees
* [x] SPHINCS+ hypertree
* [x] Integration in Botan's Pubkey Infrastructure

### CODE REVIEW

**Comment by randombit** [comment]
I noticed you defined params for Haraka as well. If you plan to support this (I am ambivalent either way on this point) please create a distinct PR for adding it.

**Comment by reneme** [comment]
> I noticed you defined params for Haraka as well. If you plan to support this (I am ambivalent either way on this point) please create a distinct PR for adding it.

It's not quite clear whether we'll add Haraka. Currently the parameters were added for completeness. Certainly, support for Haraka won't come in this first PR.

File: src/tests/test_sphincsplus_utils.cpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+namespace {
+
+std::vector<Test::Result> test_sphincsplus_address()
+   {
+   auto sha256 = [](const Botan::Sphincs_Address& adrs)
+      {
+      auto h = Botan::HashFunction::create_or_throw("SHA-256");
+      adrs.apply_to_hash(*h);
+      return h->final_stdvec();
+      };
```
Comment (reneme) [inline src/tests/test_sphincsplus_utils.cpp:26]: Add test for `apply_to_hash_compressed()`

File: src/tests/test_sphincsplus_fors.cpp [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+                                                                  sig,
+                                                                  public_seed,
+                                                                  address,
+                                                                  params,
+                                                                  *hashes);
+         result.test_is_eq("Public key from signature", pk_from_sig, pk);
+
+         if(result.tests_failed() > 0){
+            int x = 0; // Dummy
+         }
```
Comment (reneme) [inline src/tests/test_sphincsplus_fors.cpp:83]: ?

**Comment by reneme** [comment]
Closing in favor of #3549.

**Pull request closed without merging**